rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala

```diff
@@ -875,6 +935,82 @@ class DebruijnInterpreter[M[_]: Sync: Parallel: _cost](
       }
   }
 
+  private[this] val toInt: Method = new Method() {
+
+    def bigIntToLong(bi: BigInt): M[Long] =
+      if (bi > Long.MaxValue || bi < Long.MinValue)
```

Comment:
@tgrospic we already have similar code in many places for example:
https://github.com/hilltracer/rchain/blob/d6bcb0b2e7527c7f98d31877248db1b0020672f1/rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala#L1631-L1657
https://github.com/hilltracer/rchain/blob/d6bcb0b2e7527c7f98d31877248db1b0020672f1/rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala#L1601-L1629
https://github.com/hilltracer/rchain/blob/d6bcb0b2e7527c7f98d31877248db1b0020672f1/rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala#L1591-L1599
https://github.com/hilltracer/rchain/blob/d6bcb0b2e7527c7f98d31877248db1b0020672f1/rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala#L1523-L1541
Are you sure that's it is wrong? Now it works fine. And if it's wrong what we should to do with this places?

Reply:
Consider this example:
```scala
import cats.Applicative
import cats.effect.{IO, Sync}
import cats.syntax.all._

def plus(x: Int, y: Int): Int = {
  println(s"LOG: plus($x, $y)")
  x + y
}

def pureExample[F[_]: Applicative](x: Int, y: Int): F[Int] =
  plus(x, y).pure[F]

def delayExample[F[_]: Sync](x: Int, y: Int): F[Int] =
  Sync[F].delay(plus(x, y))

// Create effectful terms

val pureExampleIO: IO[Int] = pureExample[IO](1, 2)
// LOG: plus(1, 2)      <= plus already executed!!!

val delayExampleIO: IO[Int] = delayExample[IO](3, 4)
// plus in not executed!!! we are only constructing effectful value

// Evaluate effectful terms

val pureExampleRes1: Int = pureExampleIO.unsafeRunSync()
val pureExampleRes2: Int = pureExampleIO.unsafeRunSync()

val delayExampleRes1: Int = delayExampleIO.unsafeRunSync()
// LOG: plus(3, 4)
val delayExampleRes2: Int = delayExampleIO.unsafeRunSync()
// LOG: plus(3, 4)
```
So the result is correct, but the difference is **when** terms are evaluated. `F` expressions are like **pure values**, they should produce the same result (and effects) when they are evaluated and not when they are created/constructed.

This is another example where `plus` is executed even if not needed because in the pure example it's not behaving as a _pure value_.
```scala
def smartPlusWithPure[F[_]: Sync](x: Int, y: Int) = {
  val plusValue = pureExample(x, y)
  if (x == 0) y.pure[F]
  else if (y == 0) x.pure[F]
  else plusValue
}
val smartPlusWithPureRes = smartPlusWithPure[IO](1, 0).unsafeRunSync()
// LOG: plus(1, 0)      <= plus executed although not needed!!!

def smartPlusWithDelay[F[_]: Sync](x: Int, y: Int) = {
  val plusValue = delayExample(x, y)
  if (x == 0) y.pure[F]
  else if (y == 0) x.pure[F]
  else plusValue
}
val smartPlusWithPureRes = smartPlusWithDelay[IO](2, 0).unsafeRunSync()
```
It can be even worse if you replace printing to the console with some kind of exception which will be an example where the result can be different/unexpected between these examples.

So yes, existing code is wrong. It's working fine because the end result is the same, but we don't know if we will get to the edge case where this will make a difference or if it's running parts even if not needed.